src/js/reducers/conversation-reducer.js

```diff
@@ -36,22 +34,75 @@ const addMessage = (messages, message) => {
     return sortMessages([...messages, message]);
 };
 
+
+const matchMessage = (message, queryProps) => Object.keys(queryProps).every((key) => message[key] === queryProps[key]);
+
+const replaceMessage = (messages, query, newMessage) => {
+    const existingMessage = messages.find((message) => matchMessage(message, query));
+    if (!existingMessage) {
+        return messages;
+    }
+
+    const index = messages.indexOf(existingMessage);
+    return [...messages.slice(0, index), newMessage, ...messages.slice(index + 1)];
+};
+
+const isEqual = (messageA, messageB) => {
+    if (messageA._id && messageB._id && messageA._id === messageB._id) {
+        return true;
+    }
+
+    if (!messageA._id || !messageB._id) {
+        if (messageA.role === messageB.role) {
+            if (messageA.text && messageB.text && messageA.text === messageB.text) {
+                return true;
+            }
+
+            if (messageA.mediaType === messageB.mediaType && messageA.mediaUrl === messageB.mediaUrl) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+};
+
+const mergeMessages = (messagesA, messagesB) => {
+    // concat will make a union out of both arrays
+    return removeDuplicates(messagesA.concat(messagesB));
+};
+
+const removeDuplicates = (messages) => {
```

Comment:
also if you're worried about the function's efficiency then implement an object hash function for messages. Eg concat the id, role, text media type together and hash it. Then your remove duplicates can be O(n)